### PR TITLE
Add `mypy`

### PR DESCRIPTION
**Pull request by han0110**
This PR aims to add mypy for more strict type checking.

Following @ed255's suggestion in [this comment](https://github.com/appliedzkp/zkevm-specs/pull/116#issuecomment-1046723302), it introduces an interface `Expression` providing function `expr` which returns a `FQ`. In this way, we are always comparing number in field, and can avoid unsupported types comparing or integer underflow issues.

It's built upon #110 because #110 removes some unused things, then we don't bother to fix typing of them.

Co-authored-by: @ChihChengLiang
Co-authored-by: @ed255

### CODE REVIEW

**Comment by ed255** [comment]
I plan on reviewing this once https://github.com/appliedzkp/zkevm-specs/pull/110 is merged :)

File: tests/evm/test_add.py [hunk 1]
```diff
@@ -43,11 +42,11 @@ def test_add(opcode: Opcode, a: int, b: int, c: Optional[int]):
         tx_table=set(),
         bytecode_table=set(bytecode.table_assignments(randomness)),
         rw_table=set(
-            [
-                (9, RW.Read, RWTableTag.Stack, 1, 1022, 0, a, 0, 0, 0),
-                (10, RW.Read, RWTableTag.Stack, 1, 1023, 0, b, 0, 0, 0),
-                (11, RW.Write, RWTableTag.Stack, 1, 1023, 0, c, 0, 0, 0),
-            ]
+            RWDictionary(9)
```
Comment (ed255) [inline tests/evm/test_add.py:45]: This looks very nice!  I really like this `RWDictionary` class, makes tests more readable and less error prone!

**Review by ed255 — APPROVED** [review]
LGTM!

This PR is a great improvement in readability and usability! :confetti_ball:
Thanks @han0110 and @ChihChengLiang for working on this!

**Review by icemelon — APPROVED** [review]
just a nitpick, otherwise this PR looks great

**Comment by icemelon** [comment]
Could we try to merge this PR first as it's quite huge and will conflict with many open PRs? And it also introduces useful tools like `RWDictionary`

**Pull request merged**